experimental/python/databricks/bundles/core/_transform.py

```diff
@@ -0,0 +1,306 @@
+
+        kwargs = {
+            field.name: _transform_field(cls, field, value[field.name])
+            for field in cls_fields
+            if field.name in value
+        }
+
+        return cls(**kwargs)  # type:ignore
+    elif origin is list:
+        if value is None:
+            return []  # type:ignore
+        [arg] = get_args(cls)
+
+        if isinstance(value, str):
+            raise ValueError(f"Unexpected type: {cls} for '{value}'")
+
+        return [_transform(arg, item) for item in value]  # type:ignore
+    elif cls is str:
+        return str(value)  # type:ignore
+    elif cls is int:
+        return int(value)  # type:ignore
+    elif cls is bool:
+        if isinstance(value, bool):
+            return value  # type:ignore
+
+        if value in ["true", "false"]:
+            return value == "true"  # type:ignore
+
+        raise ValueError(f"Unexpected type: {cls} for '{value}'")
+    elif issubclass(origin or cls, Enum):
+        return cls(str(value))  # type:ignore
+    elif cls == NoneType and value is None:
+        return value
+    elif arg := _unwrap_dict(cls):
+        # allow to transform None to empty dict even if type doesn't allow None
+        # this is used for "create" method that has different type signature than dataclass
+        if value is None:
+            return {}  # type:ignore
+
+        [key_arg, value_arg] = arg
+
+        assert key_arg is str
+
+        if not isinstance(value, dict):
+            raise ValueError(f"Unexpected type: {cls} for '{value}'")
+
+        out = {key: _transform(value_arg, item) for key, item in value.items()}
+
+        return out  # type:ignore
+    else:
+        raise ValueError(f"Unexpected type: {cls} for {value}")
+
+
+def _unwrap_optional(tpe: type) -> Optional[type]:
+    if origin := get_origin(tpe):
+        if origin == Union or origin == UnionType:
+            args = get_args(tpe)
+
+            if len(args) == 2 and args[1] is type(None):
+                return args[0]
+            elif len(args) == 2 and args[0] is type(None):
+                return args[1]
+
+    return None
+
+
+def _unwrap_list(tpe: type) -> Optional[type]:
+    if origin := get_origin(tpe):
+        if origin is list:
+            args = get_args(tpe)
+
+            if len(args) == 1:
+                return args[0]
+
+    return None
+
+
+def _unwrap_dict(type: type) -> Optional[tuple[type, type]]:
+    if origin := get_origin(type):
+        if origin is dict:
+            args = get_args(type)
+
+            if len(args) == 2:
+                return args[0], args[1]
+
+    return None
+
+
+def _unwrap_variable(tpe: type) -> Optional[type]:
+    if origin := get_origin(tpe):
+        if origin == Variable:
+            return get_args(tpe)[0]
+
+    return None
+
+
+# from cli/libs/dyn/dynvar/ref.go
+_variable_regex = re.compile(
+    r"\$\{([a-zA-Z]+([-_]?[a-zA-Z0-9]+)*(\.[a-zA-Z]+([-_]?[a-zA-Z0-9]+)*(\[[0-9]+\])*)*(\[[0-9]+\])*)\}",
```

Comment:
The pattern has been slightly modified since the last copy (see #2203).

Reply:
Fixed in https://github.com/databricks/cli/pull/2484